File: python/vectorspace/parser.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional

from .ast import *
from .lexer import lex, Token


class ParserError(SyntaxError):
    pass


class Parser:
    def __init__(self, source: str):
        self.tokens = list(lex(source))
        self.i = 0

    @property
    def cur(self) -> Token:
        return self.tokens[self.i]

    def advance(self):
        if self.cur.kind != "EOF":
            self.i += 1

    def expect(self, kind: str) -> Token:
        if self.cur.kind != kind:
            raise ParserError(f"expected {kind}, got {self.cur.kind} at {self.cur.start}")
        tok = self.cur
        self.advance()
        return tok

    def take_ident(self) -> str:
        if self.cur.kind != "IDENT":
            raise ParserError(f"expected IDENT, got {self.cur.kind} at {self.cur.start}")
        v = self.cur.value
        self.advance()
        return v

    def take_scalar(self):
        if self.cur.kind in {"IDENT", "STRING", "INTEGER"}:
            val = self.cur.value
            self.advance()
            return val
        raise ParserError(f"expected scalar, got {self.cur.kind}")

    def take_symbolic(self) -> str:
        """
        Accept identifier-like tokens including reserved words, but reject
        punctuation and literal tokens.
        """
        if self.cur.kind in {
            "EOF",
            "INTEGER",
            "STRING",
            "(",
            ")",
            "{",
            "}",
            "[",
            "]",
            ",",
            ";",
            "=",
            ".",
        }:
            raise ParserError(f"expected identifier-like token, got {self.cur.kind} at {self.cur.start}")
        return self._consume_symbolic()

    def _consume_symbolic(self) -> str:
        value = self.cur.value if self.cur.kind == "IDENT" else self.cur.kind
        self.advance()
        return value

    def parse_program(self) -> Program:
        stmts = []
        while self.cur.kind != "EOF":
            stmts.append(self.parse_statement())
        return Program(stmts)

    def parse_statement(self):
        kind = self.cur.kind
        if kind == "vector":
            return self.parse_vector_decl()
        if kind == "wallet":
            return self.parse_wallet_decl()
        if kind == "certify":
            return self.parse_certify()
        if kind == "transfer":
            return self.parse_transfer()
        if kind == "drain":
            return self.parse_drain()
        if kind == "project":
            return self.parse_project()
        if kind == "reconstruct":
            return self.parse_reconstruct()
        if kind == "query":
            return self.parse_query()
        if kind == "record":
            return self.parse_record()
        if kind == "contract":
            return self.parse_contract()
        raise ParserError(f"unexpected token {kind} at {self.cur.start}")

    def parse_vector_decl(self):
        self.expect("vector")
        name = self.take_ident()
        self.expect(":")
        vector_type = self.take_ident()
        self.expect("=")
        self.expect("(")
        components = []
        if self.cur.kind != ")":
            components.append(self.expect("INTEGER").value)
            while self.cur.kind == ",":
                self.advance()
                components.append(self.expect("INTEGER").value)
        self.expect(")")
        self.expect(";")
        return VectorDecl(name, vector_type, [int(x) for x in components])

    def parse_wallet_decl(self):
        self.expect("wallet")
        name = self.take_ident()
        self.expect("=")
        self.expect("bind")
        self.expect("(")
        pk = self.take_scalar()
        self.expect(")")
        self.expect(";")
        return WalletDecl(name, pk)

    def parse_context(self):
        self.expect("ctx")
        self.expect("(")
        args = []
        if self.cur.kind != ")":
            args.append(self.parse_named_arg())
            while self.cur.kind == ",":
                self.advance()
                args.append(self.parse_named_arg())
        self.expect(")")
        return ContextExpr(args)

    def _is_named_arg_name_token(self) -> bool:
        """
        Accept any word-like token that can legally appear as a named argument key,
        including reserved words such as amount, policy, into, from, with, etc.
        """
        return self.cur.kind not in {
            "EOF",
            "INTEGER",
            "STRING",
            "(",
            ")",
            "{",
            "}",
            "[",
            "]",
            ",",
            ";",
            "=",
            ".",
        }

    def parse_named_arg(self):
        # Named arg form: <name> = <expr>
        # The name may be IDENT or a keyword token such as amount/policy/into/from.
        if self.cur.kind in {")", ",", "EOF"}:
            raise ParserError(f"expected named argument, got {self.cur.kind}")

        if (
            self.i + 1 < len(self.tokens)
            and self.tokens[self.i + 1].kind == "="
            and self._is_named_arg_name_token()
        ):
            name = self.cur.value if self.cur.kind == "IDENT" else self.cur.kind
            self.advance()
            self.expect("=")
            value = self.parse_expr()
            return NamedArg(name, value)

        return NamedArg(None, self.parse_expr())

    def parse_expr(self):
        if self.cur.kind == "INTEGER":
            v = int(self.cur.value)
            self.advance()
            return Expr("integer", v)

        if self.cur.kind == "STRING":
            v = self.cur.value
            self.advance()
            return Expr("string", v)

        # Any identifier-like token can begin a symbol/path/call.
        # This includes reserved words like vector, amount, policy, from, into, etc.
        if self.cur.kind not in {"INTEGER", "STRING", "(", ")", "EOF", ",", ";"}:
            value = self.take_symbolic()

            if self.cur.kind == "(":
                self.advance()
                args = []
                if self.cur.kind != ")":
                    args.append(self.parse_named_arg())
                    while self.cur.kind == ",":
                        self.advance()
                        args.append(self.parse_named_arg())
                self.expect(")")
                return Expr("call", {"callee": value, "args": args})

            parts = [value]
            while self.cur.kind == ".":
                self.advance()
                parts.append(self.take_symbolic())

            return Expr("path", parts if len(parts) > 1 else value)

        if self.cur.kind == "(":
            self.advance()
            inner = self.parse_expr()
            self.expect(")")
            return Expr("group", inner)

        raise ParserError(f"unexpected token in expression {self.cur.kind}")

    def parse_certify(self):
        self.expect("certify")
        target = self.take_ident()
        self.expect("with")
        context = self.parse_context()
        self.expect(";")
        return CertifyStmt(target, context)

    def parse_transfer(self):
        self.expect("transfer")
        source = self.take_ident()
        self.expect("to")
        destination = self.take_ident()
        self.expect("amount")
        amount = self.parse_expr()
        drain = None
        policy = None
        while self.cur.kind != ";":
            if self.cur.kind == "drain":
                self.advance()
                drain = self.parse_expr()
            elif self.cur.kind == "policy":
                self.advance()
                policy = self.parse_expr()
            else:
                raise ParserError(f"unexpected token {self.cur.kind} in transfer")
        self.expect(";")
        return TransferStmt(source, destination, amount, drain, policy)

    def parse_drain(self):
        self.expect("drain")
        target = self.take_ident()
        self.expect("by")
        amount = self.parse_expr()
        self.expect(";")
        return DrainStmt(target, amount)

    def parse_project(self):
        self.expect("project")
        source = self.take_ident()
        self.expect("into")
        environment = self.take_ident()
        self.expect("amount")
        amount = self.parse_expr()
        policy = None
        while self.cur.kind != ";":
            if self.cur.kind == "policy":
                self.advance()
                policy = self.parse_expr()
            else:
                raise ParserError(f"unexpected token {self.cur.kind} in project")
        self.expect(";")
        return ProjectStmt(source, environment, amount, policy)

    def parse_reconstruct(self):
        self.expect("reconstruct")
        target = self.take_ident()
        self.expect("from")
        projection_id = self.take_ident()
        self.expect(";")
        return ReconstructStmt(target, projection_id)

    def parse_query(self):
        self.expect("query")
        expr = self.parse_expr()
        self.expect(";")
        return QueryStmt(expr)

    def parse_record(self):
        self.expect("record")
        expr = self.parse_expr()
        self.expect(";")
        return RecordStmt(expr)

    def parse_contract(self):
        self.expect("contract")
        name = self.take_ident()
        self.expect("{")
        actions = []
        while self.cur.kind != "}":
            self.expect("action")
            action_name = self.take_ident()
            args = []
            if self.cur.kind == "(":
                self.advance()
                if self.cur.kind != ")":
                    args.append(self.parse_named_arg())
                    while self.cur.kind == ",":
                        self.advance()
                        args.append(self.parse_named_arg())
                self.expect(")")
            self.expect(";")
            actions.append(ContractAction(action_name, args))
        self.expect("}")
        return ContractDecl(name, actions)


def parse(source: str) -> Program:
    return Parser(source).parse_program()

File: src/parser.rs
use crate::ast::*;
use crate::error::{ParseError, ParseErrorKind, Span};
use crate::lexer::lex;
use crate::token::{Token, TokenKind};

#[derive(Debug, Clone)]
pub struct Parser {
    tokens: Vec<Token>,
    index: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self { tokens, index: 0 }
    }

    pub fn from_source(source: &str) -> Result<Self, ParseError> {
        Ok(Self::new(lex(source)?))
    }

    fn current(&self) -> &Token {
        &self.tokens[self.index]
    }

    fn at_end(&self) -> bool {
        self.kind_eq(&TokenKind::Eof)
    }


    fn kind_eq(&self, kind: &TokenKind) -> bool {
        std::mem::discriminant(&self.current().kind) == std::mem::discriminant(kind)
    }

    fn advance(&mut self) {
        if !self.at_end() {
            self.index += 1;
        }
    }

    fn error_here(&self, expected: &'static str) -> ParseError {
        let token = self.current();
        ParseError::new(
            Span::new(token.start, token.end),
            ParseErrorKind::UnexpectedToken {
                expected,
                found: format!("{:?}", token.kind),
            },
        )
    }

    fn expect_semi(&mut self) -> Result<(), ParseError> {
        if self.kind_eq(&TokenKind::Semi) {
            self.advance();
            Ok(())
        } else {
            Err(self.error_here("`;`"))
        }
    }

    fn expect_kind(&mut self, kind: fn(&TokenKind) -> bool, expected: &'static str) -> Result<(), ParseError> {
        if kind(&self.current().kind) {
            self.advance();
            Ok(())
        } else {
            Err(self.error_here(expected))
        }
    }

    fn take_ident(&mut self) -> Result<String, ParseError> {
        match &self.current().kind {
            TokenKind::Ident(text) => {
                let out = text.clone();
                self.advance();
                Ok(out)
            }
            other => Err(ParseError::new(
                Span::new(self.current().start, self.current().end),
                ParseErrorKind::UnexpectedToken {
                    expected: "identifier",
                    found: format!("{other:?}"),
                },
            )),
        }
    }

    fn take_string_like(&mut self) -> Result<String, ParseError> {
        match &self.current().kind {
            TokenKind::String(text) | TokenKind::Ident(text) => {
                let out = text.clone();
                self.advance();
                Ok(out)
            }
            TokenKind::Integer(v) => {
                let out = v.to_string();
                self.advance();
                Ok(out)
            }
            other => Err(ParseError::new(
                Span::new(self.current().start, self.current().end),
                ParseErrorKind::UnexpectedToken {
                    expected: "string or identifier",
                    found: format!("{other:?}"),
                },
            )),
        }
    }

    fn take_integer(&mut self) -> Result<u128, ParseError> {
        match &self.current().kind {
            TokenKind::Integer(v) => {
                let out = *v;
                self.advance();
                Ok(out)
            }
            other => Err(ParseError::new(
                Span::new(self.current().start, self.current().end),
                ParseErrorKind::UnexpectedToken {
                    expected: "integer",
                    found: format!("{other:?}"),
                },
            )),
        }
    }

    fn parse_vector_type(&mut self) -> Result<VectorType, ParseError> {
        let name = self.take_ident()?;
        VectorType::parse(&name).ok_or_else(|| {
            ParseError::new(
                Span::new(self.current().start, self.current().end),
                ParseErrorKind::Message(format!("unknown vector type `{name}`")),
            )
        })
    }

    fn parse_vector_literal(&mut self) -> Result<Vec<u128>, ParseError> {
        self.expect_kind(|k| k == &TokenKind::LParen, "`(`")?;
        let mut components = Vec::new();
        if !self.kind_eq(&TokenKind::RParen) {
            loop {
                let value = self.take_integer()?;
                components.push(value);
                if self.kind_eq(&TokenKind::Comma) {
                    self.advance();
                } else {
                    break;
                }
            }
        }
        self.expect_kind(|k| k == &TokenKind::RParen, "`)`")?;
        Ok(components)
    }

    fn parse_expr(&mut self) -> Result<Expr, ParseError> {
        self.parse_path_or_primary()
    }

    fn parse_path_or_primary(&mut self) -> Result<Expr, ParseError> {
        let mut expr = match &self.current().kind {
            TokenKind::Integer(v) => {
                let v = *v;
                self.advance();
                Expr::Integer(v)
            }
            TokenKind::String(s) => {
                let s = s.clone();
                self.advance();
                Expr::String(s)
            }
            TokenKind::Ident(name) => {
                let name = name.clone();
                self.advance();
                if self.kind_eq(&TokenKind::LParen) {
                    self.advance();
                    let args = self.parse_named_args(TokenKind::RParen)?;
                    self.expect_kind(|k| k == &TokenKind::RParen, "`)`")?;
                    Expr::Call { callee: name, args }
                } else {
                    Expr::Ident(name)
                }
            }
            TokenKind::LParen => {
                self.advance();
                let expr = self.parse_expr()?;
                self.expect_kind(|k| k == &TokenKind::RParen, "`)`")?;
                Expr::Group(Box::new(expr))
            }
            other => {
                return Err(ParseError::new(
                    Span::new(self.current().start, self.current().end),
                    ParseErrorKind::UnexpectedToken {
                        expected: "expression",
                        found: format!("{other:?}"),
                    },
                ))
            }
        };

        while self.kind_eq(&TokenKind::Dot) {
            self.advance();
            let part = self.take_ident()?;
            expr = match expr {
                Expr::Ident(first) => Expr::Path(vec![first, part]),
                Expr::Path(mut parts) => {
                    parts.push(part);
                    Expr::Path(parts)
                }
                other => {
                    return Err(ParseError::new(
                        Span::new(self.current().start, self.current().end),
                        ParseErrorKind::Message(format!("cannot apply path access to {other:?}")),
                    ))
                }
            };
        }

        Ok(expr)
    }

    fn parse_named_args(&mut self, terminator: TokenKind) -> Result<Vec<NamedArg>, ParseError> {
        let mut args = Vec::new();
        if self.kind_eq(&terminator) {
            return Ok(args);
        }
        loop {
            let arg = if let TokenKind::Ident(name) = &self.current().kind {
                let name = name.clone();
                let save = self.index;
                self.advance();
                if self.kind_eq(&TokenKind::Eq) {
                    self.advance();
                    let value = self.parse_expr()?;
                    NamedArg { name: Some(name), value }
                } else {
                    self.index = save;
                    let value = self.parse_expr()?;
                    NamedArg { name: None, value }
                }
            } else {
                let value = self.parse_expr()?;
                NamedArg { name: None, value }
            };
            args.push(arg);

            if self.kind_eq(&TokenKind::Comma) {
                self.advance();
                if self.kind_eq(&terminator) {
                    break;
                }
            } else {
                break;
            }
        }
        Ok(args)
    }

    fn parse_context_expr(&mut self) -> Result<ContextExpr, ParseError> {
        if !self.kind_eq(&TokenKind::Ctx) {
            return Err(self.error_here("ctx(...)"));
        }
        self.advance();
        self.expect_kind(|k| k == &TokenKind::LParen, "`(`")?;
        let args = self.parse_named_args(TokenKind::RParen)?;
        self.expect_kind(|k| k == &TokenKind::RParen, "`)`")?;
        Ok(ContextExpr::new(args))
    }

    fn parse_contract_action(&mut self) -> Result<ContractAction, ParseError> {
        self.expect_kind(|k| k == &TokenKind::Action, "`action`")?;
        let name = self.take_ident()?;
        let mut args = Vec::new();
        if self.kind_eq(&TokenKind::LParen) {
            self.advance();
            args = self.parse_named_args(TokenKind::RParen)?;
            self.expect_kind(|k| k == &TokenKind::RParen, "`)`")?;
        }
        self.expect_semi()?;
        Ok(ContractAction { name, args })
    }

    fn parse_statement(&mut self) -> Result<Statement, ParseError> {
        match &self.current().kind {
            TokenKind::Vector => {
                self.advance();
                let name = self.take_ident()?;
                self.expect_kind(|k| k == &TokenKind::Colon, "`:`")?;
                let vector_type = self.parse_vector_type()?;
                self.expect_kind(|k| k == &TokenKind::Eq, "`=`")?;
                let components = self.parse_vector_literal()?;
                self.expect_semi()?;
                Ok(Statement::VectorDecl(VectorDecl { name, vector_type, components }))
            }
            TokenKind::Wallet => {
                self.advance();
                let name = self.take_ident()?;
                self.expect_kind(|k| k == &TokenKind::Eq, "`=`")?;
                if !self.kind_eq(&TokenKind::Bind) {
                    return Err(self.error_here("bind(...)"));
                }
                self.advance();
                self.expect_kind(|k| k == &TokenKind::LParen, "`(`")?;
                let public_key = self.take_string_like()?;
                self.expect_kind(|k| k == &TokenKind::RParen, "`)`")?;
                self.expect_semi()?;
                Ok(Statement::WalletDecl(WalletDecl { name, public_key }))
            }
            TokenKind::Certify => {
                self.advance();
                let target = self.take_ident()?;
                if !self.kind_eq(&TokenKind::With) {
                    return Err(self.error_here("with ctx(...)"));
                }
                self.advance();
                let context = self.parse_context_expr()?;
                self.expect_semi()?;
                Ok(Statement::Certify(CertifyStmt { target, context }))
            }
            TokenKind::Transfer => {
                self.advance();
                let source = self.take_ident()?;
                if !self.kind_eq(&TokenKind::To) {
                    return Err(self.error_here("to"));
                }
                self.advance();
                let destination = self.take_ident()?;
                if !self.kind_eq(&TokenKind::Amount) {
                    return Err(self.error_here("amount"));
                }
                self.advance();
                let amount = self.parse_expr()?;
                let mut drain = None;
                let mut policy = None;
                while !self.kind_eq(&TokenKind::Semi) {
                    match &self.current().kind {
                        TokenKind::Drain => {
                            self.advance();
                            drain = Some(self.parse_expr()?);
                        }
                        TokenKind::Policy => {
                            self.advance();
                            policy = Some(self.parse_expr()?);
                        }
                        other => {
                            return Err(ParseError::new(
                                Span::new(self.current().start, self.current().end),
                                ParseErrorKind::UnexpectedToken {
                                    expected: "`drain` or `policy` or `;`",
                                    found: format!("{other:?}"),
                                },
                            ))
                        }
                    }
                }
                self.expect_semi()?;
                Ok(Statement::Transfer(TransferStmt { source, destination, amount, drain, policy }))
            }
            TokenKind::Drain => {
                self.advance();
                let target = self.take_ident()?;
                if !self.kind_eq(&TokenKind::By) {
                    return Err(self.error_here("by"));
                }
                self.advance();
                let amount = self.parse_expr()?;
                self.expect_semi()?;
                Ok(Statement::Drain(DrainStmt { target, amount }))
            }
            TokenKind::Project => {
                self.advance();
                let source = self.take_ident()?;
                if !self.kind_eq(&TokenKind::Into) {
                    return Err(self.error_here("into"));
                }
                self.advance();
                let environment = self.take_ident()?;
                if !self.kind_eq(&TokenKind::Amount) {
                    return Err(self.error_here("amount"));
                }
                self.advance();
                let amount = self.parse_expr()?;
                let mut policy = None;
                while !self.kind_eq(&TokenKind::Semi) {
                    match &self.current().kind {
                        TokenKind::Policy => {
                            self.advance();
                            policy = Some(self.parse_expr()?);
                        }
                        other => {
                            return Err(ParseError::new(
                                Span::new(self.current().start, self.current().end),
                                ParseErrorKind::UnexpectedToken {
                                    expected: "`policy` or `;`",
                                    found: format!("{other:?}"),
                                },
                            ))
                        }
                    }
                }
                self.expect_semi()?;
                Ok(Statement::Project(ProjectStmt { source, environment, amount, policy }))
            }
            TokenKind::Reconstruct => {
                self.advance();
                let target = self.take_ident()?;
                if !self.kind_eq(&TokenKind::From) {
                    return Err(self.error_here("from"));
                }
                self.advance();
                let projection_id = self.take_ident()?;
                self.expect_semi()?;
                Ok(Statement::Reconstruct(ReconstructStmt { target, projection_id }))
            }
            TokenKind::Query => {
                self.advance();
                let expr = self.parse_expr()?;
                self.expect_semi()?;
                Ok(Statement::Query(QueryStmt { expr }))
            }
            TokenKind::Record => {
                self.advance();
                let expr = self.parse_expr()?;
                self.expect_semi()?;
                Ok(Statement::Record(RecordStmt { expr }))
            }
            TokenKind::Contract => {
                self.advance();
                let name = self.take_ident()?;
                self.expect_kind(|k| k == &TokenKind::LBrace, "`{`")?;
                let mut actions = Vec::new();
                while !self.kind_eq(&TokenKind::RBrace) {
                    actions.push(self.parse_contract_action()?);
                }
                self.expect_kind(|k| k == &TokenKind::RBrace, "`}`")?;
                Ok(Statement::Contract(ContractDecl { name, actions }))
            }
            other => Err(ParseError::new(
                Span::new(self.current().start, self.current().end),
                ParseErrorKind::UnexpectedToken {
                    expected: "statement",
                    found: format!("{other:?}"),
                },
            )),
        }
    }

    pub fn parse_program(&mut self) -> Result<Program, ParseError> {
        let mut statements = Vec::new();
        while !self.at_end() {
            statements.push(self.parse_statement()?);
        }
        Ok(Program { statements })
    }
}

pub fn parse(source: &str) -> Result<Program, ParseError> {
    let mut parser = Parser::from_source(source)?;
    parser.parse_program()
}
